### PR TITLE
*[fix] Paging

### PR DESCRIPTION
**Pull request by wongpratan**
https://github.com/digi-serve/ns_app/issues/269
https://github.com/digi-serve/appbuilder_class_core/pull/175

### CODE REVIEW

**Comment by wongpratan** [comment]
@Hiro-Nakamura 

I'm sorry. They want this feature in production at the moment. 
I merged it for now.

When we will have any issue, then we can update and re-open this pull request.

**Comment by Hiro-Nakamura** [comment]
So what was the actual fix here?

File: ABDataCollectionCore.js [hunk 1]
```diff
@@ -1842,8 +1847,10 @@ module.exports = class ABDataCollectionCore extends ABMLClass {
 
          if (this.__throttleIncoming) clearTimeout(this.__throttleIncoming);
          this.__throttleIncoming = setTimeout(() => {
-            // using queuedParse() to responsively handle large datasets.
-            this.queuedParse(data.data || data, () => {
+            this.__dataCollection.load(async () => {
```
Comment (Hiro-Nakamura) [inline ABDataCollectionCore.js:1850]: One of the changes I was trying to make in the original creation of `queuedParse()` was to remove this `this.__dataCollection.load()` call.  We are no longer returning data from this fn() and instead the `queuedParse()` is actually using `this.__dataCollection.parse()` to load the data.  

Is there a reason we need to keep the `.load()` call?

**Review by Hiro-Nakamura — COMMENTED** [review]
It seems like we have basically added Promises and `await` in the process of this fix, but we haven't functionally changed how this routine works.

Why are the `await`ed Promises necessary?

**Comment by wongpratan** [comment]
> So what was the actual fix here?

The main thing is webix's `dataCollection.parse` needs to have `total_count` to support `Paging` feature.
https://docs.webix.com/desktop__plain_dynamic_loading.html#serversideresponse
https://github.com/digi-serve/appbuilder_class_core/blob/dbbc513bfe2a5547a7a045ac09b728fb8ca08623/ABDataCollectionCore.js#L1814





> It seems like we have basically added Promises and `await` in the process of this fix, but we haven't functionally changed how this routine works.
> 
> Why are the `await`ed Promises necessary?

I just updated it to support both await and callback styles. We can revert it back to Callback style if you prefer.





> One of the changes I was trying to make in the original creation of `queuedParse()` was to remove this `this.__dataCollection.load()` call. We are no longer returning data from this fn() and instead the `queuedParse()` is actually using `this.__dataCollection.parse()` to load the data.
> 
> Is there a reason we need to keep the `.load()` call?

I don't sure what exactly `.load` work but Paging feature does not work without it.
For this comment, It might need to update `total_count` value.
https://github.com/digi-serve/appbuilder_class_core/blob/50c32cc1a0cfbd6d7bb9d939e3c3d55e579715b7/ABDataCollectionCore.js#L1802

**Comment by Hiro-Nakamura** [comment]
OK, so the `.parse()` needs the full `{ .data, .pos, .total_count}` values in order to support the paging.  So that makes sense.

As long as the `fn()` sent to the `.load()` doesn't return any data, I think we can just leave it.  However, if we return data from the `fn()` then we would in effect be building the component 2x.

As for the Promises, let's go ahead and leave it as is.  unless we run into really large datasets, it should be fine.  And the promises allow us to use `await` out side the call.  However, if we notice any memory problems with large datasets, then I would consider refactoring this back to the callback only approach.  That will be more memory efficient.

**Comment by Hiro-Nakamura** [comment]
Overall, good catch on this!

**Pull request merged**